calyx-py/calyx/builder.py

```diff
@@ -898,6 +934,48 @@ def mem_latch_d2(self, mem, i, j, groupname):
             latch_grp.done = mem.done
         return latch_grp
 
+    def flatten_idx(self, dims, indices):
+        """Translate an n-dimensional index into a corresponding 1d index"""
+        assert len(dims) == len(indices)
+        i = len(indices) - 1
+        prod = 1
+        total = indices[-1]
+        while i > 0:
+            prod *= dims[i]
+            total += prod * indices[i-1]
+            i -= 1
+        return total
+    
+    def mem_load_dn(self, mem, dims, indices, reg, groupname):
+        """Inserts wiring into `self` to perform `reg := mem[i1][i2]...[in]`,
+        where `mem` is a seq_dn memory or a comb_mem_d1 memory
+        """
+        assert mem.is_seq_mem_d1() or mem.is_comb_mem_d1()
+        is_comb = mem.is_comb_mem_d1()
+        with self.group(groupname) as load_grp:
+            mem.addr0 = self.flatten_idx(dims, indices)
+            if is_comb:
+                reg.write_en = 1
+                reg.in_ = mem.read_data
+            else:
+                mem.content_en = 1
+                reg.write_en = mem.done @ 1
+                reg.in_ = mem.done @ mem.read_data
+            load_grp.done = reg.done
+        return load_grp
+
+    def mem_latch_dn(self, mem, dims, indices, groupname):
```

Comment:
I think if I'm understanding this correctly, this requires `indices` to be constants at compile-time. If so, it would be useful to add an assert to check that all the indices are in fact fact and throw an error if they are not.